text/1180971/async_afek_gafni.tex

```diff
@@ -0,0 +1,181 @@
+\subsection*{Candidate Process}
+\begin{enumerate}
+    \item while \(untraversed \ne \phi\)
+    \begin{enumerate}
+        \item choose \(l \in untraversed\)
+        \item send (\(level, id\)) along \(l\)
+        \item R: receive(\(level', id'\)) from \(l'\)
+        \item if \(id' == id\) AND \(!killed\) then
+        \begin{itemize}
+            \item \(level = level + 1\)
+            \item \(untraversed = untraversed - l\)
+        \end{itemize}
+        \item else
+        \begin{enumerate}
+            \item if \((level', id') < (level, id)\) then DISCARD message GOTO R
+            \item else 
+            \begin{itemize}
+                \item send \((level', id')\) along \(l'\)
+                \item \(killed\) = true
+                GOTO R
+            \end{itemize}
+        \end{enumerate}
+    \end{enumerate}
+    \item if \(!killed\) then \(announce(ELECTED)\)
+\end{enumerate}
+
+\subsection*{Ordinary Process}
+While (not terminated)
+\begin{enumerate}
+    \item receive(\(level', id'\)) from \(l'\)
+    \item case \(level', id'\):
+    \begin{enumerate}
+        \item \(level', id' < level, owner\_id\) DISCARD message
+        \item \(level', id' > level, owner\_id\)
+        \begin{itemize}
+            \item \(potential\_father = l'\)
+            \item \(level = level'\)
+            \item \(owner\_id = id'\)
+            \item \(if father == nil\) then \(father = potential\_father\)
+            \item send (\(level',id'\)) along \(father\)
+        \end{itemize}
+        \item \(level', id' == level, owner_id\)
+        \begin{itemize}
+            \item \(father = potential\_father\)
+            \item send (\(level',id'\)) along \(father\)
+        \end{itemize}
+    \end{enumerate}
+\end{enumerate}
+
+\section*{Messages}
+We will define some types of messages:
+\begin{itemize}
+    \item ARRIVE - the message send by candidate process to ordinary to try to capture the ordinary node. If the ordinary process of the node was never captured that means the candidate process of the node is still alive. The ARRIVE sequence will cause one of the to die.
+    \item ACCEPT\_ANSWER - the message send by the ordinary process to candidate to accept him as his father, the ordinary process is no longer candidate (if it was).
+    \item ASK - the message send by the ordinary process to candidate because he saw node with higher (level, id) than the ordinary node - he send it to the father to eliminate one node (father or sender). He waits for one of the three messages below (he doesn't process other messages in between - not including LEADER, END).
+    \item ANSWER\_ACCEPT\_CND - the message send by candidate process in response to ASK message. The candidate accepted his fate and was killed by the ASK message, he needs to inform the sender of ASK message.
+    \item ANSWER\_DENY\_CND - the message send by candidate process in response to ASK message. The candidate managed to defeat ASK message but need to response to the asker.
```

Comment:
Please change both CNDs to CANDIDATE, so that one wouldn't have to guess what it means.

Reply:
Changed CND in description and code to CANDIDATE